File: csharp/UrakawaToolkit/DataModelFactory.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Reflection;

namespace urakawa
{
	/// <summary>
	/// Factory creating the data model objects, includes creator methods for:
	/// <list type="ul">
	/// <item><see cref="Presentation"/></item>
	/// <item><see cref="metadata.MetadataFactory"/></item>
	/// <item><see cref="undo.CommandFactory"/></item>
	/// <item><see cref="undo.UndoRedoManager"/></item>
	/// <item><see cref="core.TreeNodeFactory"/></item>
	/// <item><see cref="property.PropertyFactory"/></item>
	/// <item><see cref="property.channel.ChannelFactory"/></item>
	/// <item><see cref="property.channel.ChannelsManager"/></item>
	/// <item><see cref="media.MediaFactory"/></item>
	/// <item><see cref="media.data.MediaDataFactory"/></item>
	/// <item><see cref="media.data.MediaDataManager"/></item>
	/// <item><see cref="media.data.FileDataProviderFactory"/></item>
	/// <item><see cref="media.data.FileDataProviderManager"/></item>
	/// </list>
	/// </summary>
	public class DataModelFactory
	{
		/// <summary>
		/// Creates a default <typeparamref name="T"/> object by calling <see cref="create{T}(string,string)"/> 
		/// using <c><see cref="ToolkitSettings.XUK_NS"/>:typeof(<typeparamref name="T"/>).Name</c> as Xuk QName
		/// </summary>
		/// <typeparam name="T">The object type to create</typeparam>
		/// <returns>The created <typeparamref name="T"/> instance</returns>
		private T create<T>() where T : class
		{
			return create<T>(typeof(T).Name, ToolkitSettings.XUK_NS);
		}

		/// <summary>
		/// Creates a <typeparamref name="T"/> matching a given Xuk QName
		/// </summary>
		/// <typeparam name="T">The object type to create</typeparam>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <typeparamref name="T"/> instance</returns>
		private T create<T>(string localName, string namespaceUri) where T : class
		{
			T res = null;
			if (namespaceUri==ToolkitSettings.XUK_NS && localName==typeof(T).Name)
			{
				foreach (ConstructorInfo ci in typeof(T).GetConstructors(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public))
				{
					if (ci.GetParameters().Length == 0)
					{
						res = (T)ci.Invoke(new object[0]);
					}
				}
			}
			return res;
		}

		/// <summary>
		/// Creates a <see cref="Presentation"/> of default type (that is <see cref="Presentation"/>
		/// </summary>
		/// <returns>The created <see cref="Presentation"/></returns>
		public virtual Presentation createPresentation()
		{
			return create<Presentation>();
		}

		/// <summary>
		/// Creates a <see cref="Presentation"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="Presentation"/></returns>
		public virtual Presentation createPresentation(string localName, string namespaceUri)
		{
			return create<Presentation>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="metadata.MetadataFactory"/> of default type (that is <see cref="metadata.MetadataFactory"/>
		/// </summary>
		/// <returns>The created <see cref="metadata.MetadataFactory"/></returns>
		public virtual metadata.MetadataFactory createMetadataFactory()
		{
			return create<metadata.MetadataFactory>();
		}

		/// <summary>
		/// Creates a <see cref="metadata.MetadataFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="metadata.MetadataFactory"/></returns>
		public virtual metadata.MetadataFactory createMetadataFactory(string localName, string namespaceUri)
		{
			return create<metadata.MetadataFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="undo.CommandFactory"/> of default type (that is <see cref="undo.CommandFactory"/>
		/// </summary>
		/// <returns>The created <see cref="undo.CommandFactory"/></returns>
		public virtual undo.CommandFactory createCommandFactory()
		{
			return create<undo.CommandFactory>();
		}

		/// <summary>
		/// Creates a <see cref="undo.CommandFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="undo.CommandFactory"/></returns>
		public virtual undo.CommandFactory createCommandFactory(string localName, string namespaceUri)
		{
			return create<undo.CommandFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="undo.UndoRedoManager"/> of default type (that is <see cref="undo.UndoRedoManager"/>
		/// </summary>
		/// <returns>The created <see cref="undo.UndoRedoManager"/></returns>
		public virtual undo.UndoRedoManager createUndoRedoManager()
		{
			return create<undo.UndoRedoManager>();
		}

		/// <summary>
		/// Creates a <see cref="undo.UndoRedoManager"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="undo.UndoRedoManager"/></returns>
		public virtual undo.UndoRedoManager createUndoRedoManager(string localName, string namespaceUri)
		{
			return create<undo.UndoRedoManager>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="core.TreeNodeFactory"/> of default type (that is <see cref="core.TreeNodeFactory"/>
		/// </summary>
		/// <returns>The created <see cref="core.TreeNodeFactory"/></returns>
		public virtual core.TreeNodeFactory createTreeNodeFactory()
		{
			return create<core.TreeNodeFactory>();
		}

		/// <summary>
		/// Creates a <see cref="core.TreeNodeFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="core.TreeNodeFactory"/></returns>
		public virtual core.TreeNodeFactory createTreeNodeFactory(string localName, string namespaceUri)
		{
			return create<core.TreeNodeFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="property.PropertyFactory"/> of default type (that is <see cref="property.PropertyFactory"/>
		/// </summary>
		/// <returns>The created <see cref="property.PropertyFactory"/></returns>
		public virtual property.PropertyFactory createPropertyFactory()
		{
			return create<property.PropertyFactory>();
		}

		/// <summary>
		/// Creates a <see cref="property.PropertyFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="property.PropertyFactory"/></returns>
		public virtual property.PropertyFactory createPropertyFactory(string localName, string namespaceUri)
		{
			return create<property.PropertyFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="property.channel.ChannelFactory"/> of default type (that is <see cref="property.channel.ChannelFactory"/>
		/// </summary>
		/// <returns>The created <see cref="property.channel.ChannelFactory"/></returns>
		public virtual property.channel.ChannelFactory createChannelFactory()
		{
			return create<property.channel.ChannelFactory>();
		}

		/// <summary>
		/// Creates a <see cref="property.channel.ChannelFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="property.channel.ChannelFactory"/></returns>
		public virtual property.channel.ChannelFactory createChannelFactory(string localName, string namespaceUri)
		{
			return create<property.channel.ChannelFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="property.channel.ChannelsManager"/> of default type (that is <see cref="property.channel.ChannelsManager"/>
		/// </summary>
		/// <returns>The created <see cref="property.channel.ChannelsManager"/></returns>
		public virtual property.channel.ChannelsManager createChannelsManager()
		{
			return create<property.channel.ChannelsManager>();
		}

		/// <summary>
		/// Creates a <see cref="property.channel.ChannelsManager"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="property.channel.ChannelsManager"/></returns>
		public virtual property.channel.ChannelsManager createChannelsManager(string localName, string namespaceUri)
		{
			return create<property.channel.ChannelsManager>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="media.IMediaFactory"/> of default type (that is <see cref="media.MediaFactory"/>
		/// </summary>
		/// <returns>The created <see cref="media.MediaFactory"/></returns>
		public virtual media.IMediaFactory createMediaFactory()
		{
			return create<media.MediaFactory>();
		}

		/// <summary>
		/// Creates a <see cref="media.IMediaFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="media.IMediaFactory"/></returns>
		public virtual media.IMediaFactory createMediaFactory(string localName, string namespaceUri)
		{
			return create<media.MediaFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="media.data.MediaDataFactory"/> of default type (that is <see cref="media.data.MediaDataFactory"/>
		/// </summary>
		/// <returns>The created <see cref="media.data.MediaDataFactory"/></returns>
		public virtual media.data.MediaDataFactory createMediaDataFactory()
		{
			return create<media.data.MediaDataFactory>();
		}

		/// <summary>
		/// Creates a <see cref="media.data.MediaDataFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="media.data.MediaDataFactory"/></returns>
		public virtual media.data.MediaDataFactory createMediaDataFactory(string localName, string namespaceUri)
		{
			return create<media.data.MediaDataFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="media.data.MediaDataManager"/> of default type (that is <see cref="media.data.MediaDataManager"/>
		/// </summary>
		/// <returns>The created <see cref="media.data.MediaDataManager"/></returns>
		public virtual media.data.MediaDataManager createMediaDataManager()
		{
			return create<media.data.MediaDataManager>();
		}

		/// <summary>
		/// Creates a <see cref="media.data.MediaDataManager"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="media.data.MediaDataManager"/></returns>
		public virtual media.data.MediaDataManager createMediaDataManager(string localName, string namespaceUri)
		{
			return create<media.data.MediaDataManager>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="media.data.IDataProviderFactory"/> of default type (that is <see cref="media.data.FileDataProviderFactory"/>
		/// </summary>
		/// <returns>The created <see cref="media.data.IDataProviderFactory"/></returns>
		public virtual media.data.IDataProviderFactory createDataProviderFactory()
		{
			return create<media.data.FileDataProviderFactory>();
		}

		/// <summary>
		/// Creates a <see cref="media.data.IDataProviderFactory"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="media.data.IDataProviderFactory"/></returns>
		public virtual media.data.IDataProviderFactory createDataProviderFactory(string localName, string namespaceUri)
		{
			return create<media.data.FileDataProviderFactory>(localName, namespaceUri);
		}

		/// <summary>
		/// Creates a <see cref="media.data.IDataProviderManager"/> of default type (that is <see cref="media.data.FileDataProviderManager"/>
		/// </summary>
		/// <returns>The created <see cref="media.data.IDataProviderManager"/></returns>
		public virtual media.data.IDataProviderManager createDataProviderManager()
		{
			return create<media.data.FileDataProviderManager>();
		}

		/// <summary>
		/// Creates a <see cref="media.data.IDataProviderManager"/> of type matching a given Xuk QName
		/// </summary>
		/// <param name="localName">The local name part of the given Xuk QName</param>
		/// <param name="namespaceUri">The namespace uri part of the given Xuk QName</param>
		/// <returns>The created <see cref="media.data.IDataProviderManager"/></returns>
		public virtual media.data.IDataProviderManager createDataProviderManager(string localName, string namespaceUri)
		{
			return create<media.data.FileDataProviderManager>(localName, namespaceUri);
		}
	}
}

File: trunk/java/core/src/org/daisy/urakawa/DataModelFactory.java
package org.daisy.urakawa;

import org.daisy.urakawa.command.ICommandFactory;
import org.daisy.urakawa.command.CommandFactory;
import org.daisy.urakawa.core.ITreeNodeFactory;
import org.daisy.urakawa.core.TreeNodeFactory;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.media.IMediaFactory;
import org.daisy.urakawa.media.MediaFactory;
import org.daisy.urakawa.media.data.IDataProviderFactory;
import org.daisy.urakawa.media.data.IDataProviderManager;
import org.daisy.urakawa.media.data.DataProviderFactory;
import org.daisy.urakawa.media.data.DataProviderManager;
import org.daisy.urakawa.media.data.IMediaDataFactory;
import org.daisy.urakawa.media.data.MediaDataFactory;
import org.daisy.urakawa.media.data.IMediaDataManager;
import org.daisy.urakawa.media.data.MediaDataManager;
import org.daisy.urakawa.metadata.IMetadataFactory;
import org.daisy.urakawa.metadata.MetadataFactory;
import org.daisy.urakawa.property.IPropertyFactory;
import org.daisy.urakawa.property.PropertyFactory;
import org.daisy.urakawa.property.channel.IChannelFactory;
import org.daisy.urakawa.property.channel.ChannelFactory;
import org.daisy.urakawa.property.channel.IChannelsManager;
import org.daisy.urakawa.property.channel.ChannelsManager;
import org.daisy.urakawa.undo.IUndoRedoManager;
import org.daisy.urakawa.undo.UndoRedoManager;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * Reference implementation of the interface.
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 */
public final class DataModelFactory implements IDataModelFactory {
	/**
	 * TODO: Check that this instantiation mechanism actually works in Java 1.5
	 * 
	 * @param <T>
	 * @param klass
	 * @param xukLocalName
	 * @param xukNamespaceUri
	 * @return the created instance
	 * @throws MethodParameterIsNullException
	 * @throws MethodParameterIsEmptyStringException
	 */
	private <T> T create(Class<T> klass, String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		if (xukLocalName == null || xukNamespaceUri == null) {
			throw new MethodParameterIsNullException();
		}
		if (xukLocalName.length() == 0) {
			throw new MethodParameterIsEmptyStringException();
		}
		String str = klass.getSimpleName();
		
		if (str != xukLocalName || xukNamespaceUri != IXukAble.XUK_NS) {
			return null;
		}
		try {
			return klass.newInstance();
		} catch (InstantiationException e) {
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		}
		return null;
	}

	public IChannelFactory createChannelFactory() {
		return new ChannelFactory();
	}

	public IChannelFactory createChannelFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(ChannelFactory.class, xukLocalName, xukNamespaceUri);
	}

	public IChannelsManager createChannelsManager() {
		return new ChannelsManager();
	}

	public IChannelsManager createChannelsManager(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(ChannelsManager.class, xukLocalName, xukNamespaceUri);
	}

	public ICommandFactory createCommandFactory() {
		return new CommandFactory();
	}

	public ICommandFactory createCommandFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(CommandFactory.class, xukLocalName, xukNamespaceUri);
	}

	public IDataProviderFactory createDataProviderFactory() {
		return new DataProviderFactory();
	}

	public IDataProviderFactory createDataProviderFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(DataProviderFactory.class, xukLocalName,
				xukNamespaceUri);
	}

	public IDataProviderManager createDataProviderManager() {
		return new DataProviderManager();
	}

	public IDataProviderManager createDataProviderManager(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(DataProviderManager.class, xukLocalName,
				xukNamespaceUri);
	}

	public IMediaDataFactory createMediaDataFactory() {
		return new MediaDataFactory();
	}

	public IMediaDataFactory createMediaDataFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(MediaDataFactory.class, xukLocalName, xukNamespaceUri);
	}

	public IMediaDataManager createMediaDataManager() {
		return new MediaDataManager();
	}

	public IMediaDataManager createMediaDataManager(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(MediaDataManager.class, xukLocalName, xukNamespaceUri);
	}

	public IMediaFactory createMediaFactory() {
		return new MediaFactory();
	}

	public IMediaFactory createMediaFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(MediaFactory.class, xukLocalName, xukNamespaceUri);
	}

	public IMetadataFactory createMetadataFactory() {
		return new MetadataFactory();
	}

	public IMetadataFactory createMetadataFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(MetadataFactory.class, xukLocalName, xukNamespaceUri);
	}

	public IPresentation createPresentation() {
		return new Presentation();
	}

	public IPresentation createPresentation(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(Presentation.class, xukLocalName, xukNamespaceUri);
	}

	public IPropertyFactory createPropertyFactory() {
		return new PropertyFactory();
	}

	public IPropertyFactory createPropertyFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(PropertyFactory.class, xukLocalName, xukNamespaceUri);
	}

	public ITreeNodeFactory createTreeNodeFactory() {
		return new TreeNodeFactory();
	}

	public ITreeNodeFactory createTreeNodeFactory(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(TreeNodeFactory.class, xukLocalName, xukNamespaceUri);
	}

	public IUndoRedoManager createUndoRedoManager() {
		return new UndoRedoManager();
	}

	public IUndoRedoManager createUndoRedoManager(String xukLocalName,
			String xukNamespaceUri) throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		return create(UndoRedoManager.class, xukLocalName, xukNamespaceUri);
	}
}
